### ISSUE TITLE
osd prepare pod goes into reconcile loop

### ISSUE THREAD

**Issue by umangachapagain** [rook/rook#3029]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
osd prepare pod goes into reconcile loop even if there is no change in OSD or Disks.

**Expected behavior:**
osd prepare pod should go into completed state once it's job is done.

**How to reproduce it (minimal and precise):**
1. Deploy an OCP 4 cluster.
2. Deploy Rook + Ceph as documented.

**Environment**:
* OS (e.g. from /etc/os-release): CentOS Linux 7 (Core)
* Kernel (e.g. `uname -a`): Linux 3.10.0-957.10.1.el7.x86_64
* Rook version (use `rook version` inside of a Rook Pod): v0.9.0-465.g5f6de03
* oc version (use `oc version`): v4.1.0
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): OpenShift
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)): 
 ```
health: HEALTH_WARN
            noscrub,nodeep-scrub flag(s) set
```

Operator Pod Logs: http://pastefile.fr/62a2aafcecf083cfc35c26ecde9cdfd9

**Comment by umangachapagain** [comment]
@leseb @travisn

**Comment by travisn** [comment]
Rook is detecting a device change and triggering the orchestration. @umangachapagain Could you look at the discover pod logs and see if there are any more messages about the devices that are changing? Here is the message from the operator log found many times:
```
Running orchestration for namespace openshift-storage after device change
```

**Comment by umangachapagain** [comment]
Discover Pod logs:
http://pastefile.fr/6b286c6a6e7e4d4485a26a77897f12e2
http://pastefile.fr/53ac838b5ae4e64b6d22d9b56ef6def9
http://pastefile.fr/c6c1100e27499a29ae796d1029904cc3

**Comment by noahdesu** [comment]
we should probably add blacklist filters somewhere they can be configured. in the crd yaml file?

**Comment by travisn** [comment]
@noahdesu The discover pod can actually feed multiple ceph clusters (although the common case is only a single cluster), so this maps better to a setting in the operator, for example as an environment var in operator.yaml. What about a comma-separated blacklist var or else a regex that would get passed down to the discover pod? Or what about the opposite approach where it's a white list? Then the default setting could be very conservative so that we don't trigger too many orchestrations unintentionally.

**Comment by noahdesu** [comment]
Update from an out-of-band discussion with @travisn 

The discovery logs show that orchestrations are being kicked off by the detection of `rbd` devices. Like device mapper devices it makes sense to blacklist `rbd` devices from detection for the purpose of creating new osds. This is resolved here https://github.com/rook/rook/pull/3032.

We've also added in the previous PR an operator configuration option that completely disables hotplug orchestration in case things get stuck in a loop again.

Finally, the root cause of the reconciliation loop is still unknown. The operator log shows new orchestrations being run every few minutes. AFAICT the _only_ source of this trigger is the discovery mechanism. However, those logs show the trigger as relatively quiet (e.g. mostly once per hour).

@umangachapagain are there more discovery logs? in particular, the logs posted above mostly do not overlap in time with the operator log. would there be other discovery pods whose logs were omitted (or perhaps trimmed due to a restart?).

@liewegas

**Comment by travisn** [comment]
This was fixed by #3032
